src/v/ssx/watchdog.h

```diff
@@ -55,18 +54,56 @@ class watchdog {
     ///
     /// \param timeout defines time interval after which the watchdog will be
     ///                triggered
-    /// \param deadline_reached is a callback that will be called when
-    ///                         the watchdog is triggered
+    /// \param callback is a callback that will be called when
+    ///                 the watchdog is triggered
     /// \note The callback may outlive the watchdog
     /// instance. To prevent lifetime issues one could use external
     /// synchronization (hold a gate in the callback and close the gate outside
     /// of the callback) or alternatively (and preferably) the callback should
     /// be very simple, for instance, it should just log an error message.
     watchdog(
       seastar::lowres_clock::duration timeout,
-      seastar::noncopyable_function<void()> deadline_reached) {
-        start_waiting(timeout, std::move(deadline_reached));
+      seastar::noncopyable_function<void()> callback,
+      double growth_factor = 0,
+      seastar::lowres_clock::duration max_timeout
+      = seastar::lowres_clock::duration::max()) {
+        if (growth_factor == 0) {
+            start_waiting(timeout, std::move(callback));
+        } else {
+            ssx::background = ssx::ignore_shutdown_exceptions(seastar::do_with(
+              timeout,
+              std::move(callback),
+              growth_factor,
+              max_timeout,
+              [this](
+                seastar::lowres_clock::duration& timeout,
+                seastar::noncopyable_function<void()>& cb,
+                double growth_factor,
+                seastar::lowres_clock::duration max_timeout) {
+                  return seastar::sleep_abortable(timeout, _as)
+                    .then([&cb] { cb(); })
+                    .then([this, &timeout, &cb, growth_factor, max_timeout] {
+                        return seastar::repeat(
+                          [this, &timeout, &cb, growth_factor, max_timeout] {
+                              timeout = std::chrono::duration_cast<
+                                seastar::lowres_clock::duration>(
+                                timeout * growth_factor);
+
+                              if (timeout > max_timeout) {
+                                  timeout = max_timeout;
+                              }
+
+                              return seastar::sleep_abortable(timeout, _as)
```

Comment:
i think this has a subtle use-after-free risk. if the watchdog timer is on the stack (that sees common from what i've sene), the first sleep expires, then a scheduling point is introduce after one of the subsequent `.then` continuations, then the caller returns and destroys the slack, this background fiber will keep running and call sleep_abortable again but now with a freed abort source?

Reply:
Sounds plausible.